### PR TITLE
Add a tagging policy

### LINKED ISSUE #72: Publish all images with extra tags which includes the commit hash
If users want to bind their running version to defined by a commit. These tags would increase the high confident in running Janugraph than just version tags. 

Issue: We could introduce breaking changes in the docker images without a bump of the JanusGraph version.

### PR DESCRIPTION
**Pull request by farodin91**
Fixes #72

* Add a tagging policy to the readme.md
* Tag all images with extra tag which contains a short commit hash
* Use fullpath to push images docker.io/janusgraph/janusgraph instead of janusgraph/janusgraph
* Use fullpath in all examples
* Quote most variables in build-images.sh and push-images.sh
* Add changelog.md
 
Signed-off-by: Jan Jansen <jan.jansen@gdata.de>

### CODE REVIEW

File: push-images.sh [hunk 1]
```diff
@@ -23,16 +23,22 @@ versions=($(ls -d [0-9]*))
 # get the last element of sorted version folders
 latest_version=${versions[${#versions[@]}-1]}
 
+REVISION=$(git rev-parse --short HEAD)
```
Comment (mbrukman) [inline push-images.sh:26]: Quote the RHS of the assignment:

```
REVISION="$(git rev-parse --short HEAD)"
```

File: README.md [hunk 2]
```diff
@@ -279,6 +279,17 @@ janusgraph-mount | chown: changing ownership of '/etc/opt/janusgraph/janusgraph.
...
 ```
 
+## Docker Tagging Policy
+
+We are ensure following tagging policy for our Docker images:
+
+| Tag            | Support level |
+| -------------- | ------------- |
+| latest         | latest JanusGraph release; no breaking changes guarantees; base layer upgrades, e.g. operating system or java; |
+| 0.x            | newest patch level version of JanusGraph; expect breaking changes; base layer upgrades, e.g. operating system or java; |
```
Comment (mbrukman) [inline README.md:289]: What does it mean to expect breaking changes? If someone asks for JanusGraph 0.5, the only thing that can change is that they will get 0.5.z versions for some z, which is the latest patch-level version (as you said), but that should be backwards-compatible, no?

Other than OS / Java updates, what possible breaking changes would there be?

Comment (mbrukman) [inline README.md:289]: s/patch level/patch-level/ ?

**Review by mbrukman — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (farodin91) [reply README.md:289]: In the normal case this is true, but we already had some breaking changes in patch level versions: https://docs.janusgraph.org/changelog/#tinkerpop-upgrade-from-341-to-344

File: README.md [hunk 3]
```diff
@@ -279,6 +279,17 @@ janusgraph-mount | chown: changing ownership of '/etc/opt/janusgraph/janusgraph.
...
+| 0.x.x          | defined JanusGraph version; breaking changes are only in this repo; base layer upgrades, e.g. operating system or java; |
```
[the remark this answers is not in the document]
Comment (farodin91) [reply README.md:290]: We handle this code independent from our mainly, properly have to introduce minor changes.

**Comment by farodin91** [comment]
@mbrukman I reworked the table a bit more. It contains now information about os and java version

**Comment by farodin91** [comment]
@mbrukman Would you like to check it again?

File: 0.3/Dockerfile [hunk 4]
```diff
@@ -38,7 +38,7 @@ RUN curl -fSL https://github.com/JanusGraph/janusgraph/releases/download/v${JANU
 COPY conf/janusgraph-berkeleyje-lucene-server.properties conf/log4j-server.properties ${JANUS_HOME}/conf/gremlin-server/
 COPY scripts/remote-connect.groovy ${JANUS_HOME}/scripts/
 
-FROM openjdk:8-jdk
+FROM openjdk:8-jdk-buster
```
Comment (mbrukman) [inline 0.3/Dockerfile:41]: This change which is modifying the base image seems orthogonal to the overall change described in this PR which is adding tags (and is not mentioned in the PR description). Is the current `openjdk:8-jdk` equivalent to `openjdk:8-jdk-buster` and this is just making it explicit? Or is the goal to choose a more specific (and hence more stable) base image before instituting the tagging policy?

Also, if we're re-considering the base images, what about considering:

* using `slim-buster` instead of `buster` which is smaller
* using `jre` instead of `jdk` since it's smaller

Bottom line: would it make sense to make the base image more explicit and/or smaller, and if you want to make that change _before_ the tagging process which codifies stability guarantees, maybe it makes sense to change the base images as a separate change?

**Review by mbrukman — CHANGES_REQUESTED** [review]

File: 0.3/Dockerfile (hunk 4 above)
Comment (farodin91) [reply 0.3/Dockerfile:41]: > this is just making it explicit? 

Yes

My plan is to make it more specific in follow up pull request. I could do it before. 

I have two ideas todo. 
- Start image as JanusGraph user
- Use jre slim image #75

After this pr is merged we can search for a solution to prepare breaking changes.

**Review by mbrukman — APPROVED** [review]
LGTM, thanks for the explanation.
